Introduce Bazel Steward - a tool for keeping dependencies up to date in Bazel

lukaszwawrzyk opened the pull request:
Hello everyone,

As one of developers I am submitting this pull request to introduce our new tool for Bazel repositories. [Bazel Steward](https://github.com/VirtusLab/bazel-steward) simplifies the process of checking and updating dependencies in your Bazel projects, making it more efficient to keep them up-to-date.
We hope that this tool will be useful to the Bazel community, and we look forward to your feedback.

This pull request integrates Bazel Steward through Github Actions which is currently the easiest way to do this. It will run every day at 12 and create new PRs or resolve conflicts on existing if necessary. You can preview how it looks in [a fork](https://github.com/lukaszwawrzyk/batfish/pulls) that I used to test it. For more details, you can check [the project readme](https://github.com/VirtusLab/bazel-steward#readme). I configured a limit to keep max 10 Bazel Steward PRs at the same time, as otherwise about 40 PRs for outdated versions would be created which might be overwhelming.

Bazel Steward is able to correctly update all your maven dependencies, version of Bazel itself and most of Bazel rules.

We hope that Bazel Steward will make it easier for you to manage dependencies. If you encounter any issues or have any feedback, please don't hesitate to reach out to us. Thank you!

dhalperi commented:
Thanks @lukaszwawrzyk - this is really interesting. I have some questions:

1. We'd like to update groups (such as all the Jackson libraries, or all the jersey libraries) in tandem, rather than across many PRs per library (https://github.com/lukaszwawrzyk/batfish/pull/5). Is that supported?
2. Does the work upgrading rulesets such as rules_jvm_external also re-pin the dependencies? How did/does it discover all the install files to use? https://github.com/lukaszwawrzyk/batfish/pull/3

lukaszwawrzyk commented:
> Thanks @lukaszwawrzyk - this is really interesting. I have some questions:
> 
> 1. We'd like to update groups (such as all the Jackson libraries, or all the jersey libraries) in tandem, rather than across many PRs per library ([Updated com.fasterxml.jackson.core:jackson-annotations to 2.15.0 lukaszwawrzyk/batfish#5](https://github.com/lukaszwawrzyk/batfish/pull/5)). Is that supported?

Currently grouping is not supported, but we designed Bazel Steward with this feature in mind. I think we should be able to add it within this month. Do you consider it a blocking issue?

> 2. Does the work upgrading rulesets such as rules_jvm_external also re-pin the dependencies? How did/does it discover all the install files to use? [Updated rules_jvm_external to 5.2 lukaszwawrzyk/batfish#3](https://github.com/lukaszwawrzyk/batfish/pull/3)

Check the `bazel-steward.yaml` in this PR and any maven PR in the fork. Pinning is applied after bumping based on the config. To repin dependencies after rules update, config should be extended with:

```
post-update-hooks:
  - dependencies: rules_jvm_external
    commands:
      - "bazel run @unpinned_maven//:pin"
    files-to-commit:
      - "maven_install.json"
    run-for: commit
```

I am not sure what do you mean by "detect install files to use". The json files are not detected, you need to list them in files to commit (all files that will change after invoking the commands; globs supported).
We detect available maven dependencies through bazel, and it only works with the default `maven` workspace. If you call maven_install multiple times with different names, we don't support that yet. Let me know if this is important in your use case.

lukaszwawrzyk commented:
I deleted the fork, because I wanted to create PRs from scratch for demonstration so this PR got closed. I plan to recreate it properly if you will be interested.
I implemented grouping for libraries that share version, like jackson.

You can preview the config I used here: https://github.com/lukaszwawrzyk/batfish/blob/master/bazel-steward.yaml

And example of grouped PR: https://github.com/lukaszwawrzyk/batfish/pull/18/files

I lifted the limit of open PRs, so now all potential PRs that can be open, are open, so you can investigate it.

There is no PR for rules_jvm_external because post commit hook failed as pinning produced no diff. I will fix that one soon.

lukaszwawrzyk commented:
@dhalperi Hi! Have you had chance to test it out?

lukaszwawrzyk commented:
@dhalperi kind ping :p

lukaszwawrzyk commented:
@dhalperi The demo can be viewed in the fork. Do you mean I should recreate a PR to this repo? I can do this easily, just let me know if I understand correctly.
